### PR TITLE
 Rework markdown TOC parser

### PR DESCRIPTION
**Pull request by yufeih**
- Use `foreach` over index to make it more readable
- Lift inner function out because these inner functions captures local variable
- Change `invalid-toc-syntax` error message to not depend on source content.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/4813)

### CODE REVIEW

File: src/docfx/build/toc/parser/MarkdownTocMarkup.cs [hunk 1]
```diff
@@ -45,141 +43,111 @@ public static (List<Error> errors, TableOfContentsModel model) LoadMdTocModel(st
...
-            tocModel.Items = ConvertTo(tocContent, file.FilePath, headingBlocks.ToArray(), errors).children;
+            var tree = BuildTree(errors, file.FilePath, headingBlocks);
+
+            var tocModel = new TableOfContentsModel { Metadata = tocMetadata, Items = tree.Items };
 
             return (errors, tocModel);
         }
 
-        private static (List<TableOfContentsItem> children, int count) ConvertTo(string tocContent, string filePath, HeadingBlock[] headingBlocks, List<Error> errors, int startIndex = 0)
+        private static TableOfContentsItem BuildTree(List<Error> errors, string filePath, List<HeadingBlock> blocks)
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/MarkdownTocMarkup.cs]: >TableOfContentsItem [](start = 23, length = 19)

List<TableOfContentsItem>

File: src/docfx/build/toc/parser/MarkdownTocMarkup.cs [hunk 2]
```diff
@@ -45,141 +43,111 @@ public static (List<Error> errors, TableOfContentsModel model) LoadMdTocModel(st
...
 
-                    distance++;
+                stack.Push((currentLevel, currentItem));
+            }
 
-                    if (headingBlocks[j].Level - currentLevel == 1)
-                    {
-                        break;
-                    }
+            return result;
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/MarkdownTocMarkup.cs]: >result [](start = 19, length = 6)

result.Items

File: src/docfx/build/toc/parser/MarkdownTocMarkup.cs [hunk 3]
```diff
@@ -45,141 +43,111 @@ public static (List<Error> errors, TableOfContentsModel model) LoadMdTocModel(st
...
+                if (currentItem == null)
                 {
                     continue;
                 }
 
-                items.Add(item);
-                var currentLevel = headingBlocks[i].Level;
-                var (nextLevelDistance, skipped) = GetNextLevelDistance(i, headingBlocks[i].Level);
-                if (nextLevelDistance > 0)
+                while (stack.TryPeek(out top) && top.level >= currentLevel)
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/MarkdownTocMarkup.cs]: >                 [](start = 0, length = 16)

Try some cases like:
##### 1
#### 2
### 3

Seems the skip level error will not be reported

File: src/docfx/build/toc/parser/MarkdownTocMarkup.cs [hunk 4]
```diff
@@ -45,141 +43,123 @@ public static (List<Error> errors, TableOfContentsModel model) LoadMdTocModel(st
...
-                        break;
-                    }
-
-                    distance++;
-
-                    if (headingBlocks[j].Level - currentLevel == 1)
+                    if (rootLevel < 0)
                     {
-                        break;
+                        rootLevel = currentLevel - 1;
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/MarkdownTocMarkup.cs]: really strange here, would you turn skip-level into warning in your test case?

File: src/docfx/build/toc/parser/MarkdownTocMarkup.cs [hunk 5]
```diff
@@ -45,141 +43,123 @@ public static (List<Error> errors, TableOfContentsModel model) LoadMdTocModel(st
...
                     }
-
-                    if (reported++ == 0)
-                        errors.Add(Errors.InvalidTocLevel(headingBlocks[j].ToSourceInfo(file: filePath), currentLevel, headingBlocks[j].Level));
+                    parent.level = rootLevel;
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/MarkdownTocMarkup.cs]: it's not immutable

Comment (yufeih) [reply src/docfx/build/toc/parser/MarkdownTocMarkup.cs]: de-mystified

File: docs/specs/toc.yml [hunk 6]
```diff
@@ -1152,7 +1152,7 @@ inputs:
 outputs:
   docs/test.json:
   .errors.log: |
-    ["error","invalid-toc-syntax","The toc syntax '[bad1]()\n#[bad2](test.md)\n>_<\n#bad4' is invalid, the opening sequence of # characters must be followed by a space or by the end of line. Refer to [ATX heading](https://spec.commonmark.org/0.28/#atx-heading) to fix it","docs/TOC.md",2,1]
+    ["error","invalid-toc-syntax","The toc syntax is invalid, *",2,1]
```
Comment (OsmondJiang) [inline docs/specs/toc.yml]: what's that?

**Pull request merged**